plutus-playground-lib/src/Playground/Types.hs

```diff
@@ -45,23 +50,30 @@ adaCurrency =
         }
 
 --------------------------------------------------------------------------------
-newtype Fn =
-    Fn Text
+newtype EndpointName =
+    EndpointName Text
     deriving (Eq, Show, Generic, TH.Lift)
     deriving newtype (ToJSON, FromJSON)
 
-data Expression
-    = Action
-          { function  :: Fn
-          , wallet    :: Wallet
-          , arguments :: [JSON.Value]
-          }
-    | Wait
+instance Newtype EndpointName
+
+data Expression (schema :: Row *)
+    = AddBlocks
           { blocks :: Int
           }
+    | PayToWallet
+          { source :: Wallet
+          , destination :: Wallet
+          , value :: V.Value
+          }
+    | CallEndpoint
+          { endpointName :: EndpointName
+          , wallet :: Wallet
+          , arguments :: JSON.Value
```

Comment:
Perhaps `arguments :: Var (Input schema)` ? or maybe `arguments :: Var (Input (schema .\\ BlockchainActions))`. Or `Data.Row.Extras.JsonVar (Input (schema .\\ BlockchainActions))` to get the JSON instances

Reply:
The really thorny problem with this whole thing is this:

* The user sends us their contract and the function calls (with their arguments), as JSON. 
* We decode that JSON payload and compile the contract.
* Compilation gives us back the schema for the contract.
* And it's only _then_ that we know what type to decode the function calls as. 

So we're stuck receiving a JSON payload that we can't fully decode in one pass.

We could, I suppose, have a second datatype that decodes the same JSON but fewer fields. I'm not sure that would come out simpler though... 🤔